### PR TITLE
dqfd, noisynet + PER fix

### PR DESCRIPTION
**Pull request by jakegrigsby**
Key changes:

dqn.py: Adds my attempt at Deep Q Learning from Demonstrations.

memory.py: Adds memory for dqfd, which is a buffer that can sample expert demo data without overwriting it. That includes a custom RingBuffer.

layers.py: New file to hold custom keras layers for our models. adds NoisyNetDense, modeled after [this paper](https://arxiv.org/abs/1706.10295). Also includes NAFLayer, moved over from dqn.py

noisynet_pdd_dqn_atari.py: new example script for training (and testing) a NoisyNet Prioritized Dueling Double DQN. This is separated from the existing dqn_atari example because it requires changes to the keras model itself.

dqfd_atari.py: example that highlights the differences in setting up dqfd vs. dqn.

callbacks.py: adds the option to dump the output of TrainEpisodeLogger into a txt file. Totally optional, but useful if you're used to watching that log output during training and want to see the same stats after completion.

Testing:

noisynet_10m_reward_curve

This is the learning curve for Noisy PDD DQN on the atari game Gopher. Video of its gameplay [here](https://drive.google.com/open?id=1pCLwEi1aUw0jemQMh9PqcxHsGDk1AtER).

Here is a video of the new noisynet example script playing Boxing, and winning 100-3: [click here](https://drive.google.com/file/d/18crFwEWgTLVbw74ERni2ZrjEtmY609Hl/view?usp=sharing)

### CODE REVIEW

**Comment by jakegrigsby** [comment]
Update: dueling dqn was chopping off the last layer and building out the dueling architecture internally, meaning that models defined with 2 noisy layers were really only using one, and models with one layer had no exploration at all. This has been fixed but will probably be part of an upcoming DQfD PR, because it's in both algorithms.

**Comment by ulrikborjesson** [comment]
The n-step dqn implementation seems incorrect. The point of n-step is to look ahead and use future rewards to better approximate future discounted rewards. In this line ` Rs = (reward_batch **(self.n_step)) + discounted_reward_batch` you raise the reward to n and this is a misinterpretation of the equation for n-step dqn. The exponent is not an actual exponent in the origial algorithm, it is a notation. See this link for example: http://www.incompleteideas.net/book/7/node2.html

To implement n-step we would need to look ahead at future rewards up to n-steps in the future. This can easily be solved by sampling between the first index in the ringbuffers up to the max length - n. When there's a new episode there's no need to look further ahead and you can simply use those or that specific rewards(s) as the target.

**Comment by jakegrigsby** [comment]
Don't have time to work on that right now so if you want to do it that'd be great or I'll get it done this weekend or early next week.

**Comment by jakegrigsby** [comment]
Just added a pretty heavy update and edited the first comment to reflect the changes. DQfD's custom RingBuffer should solve the indexing issue discussed on my last PR. I gave the corrected n-step a shot within DQfD, but left it out of dqn (for now). Also includes the noisynet dueling change I was talking about.

**Comment by ulrikborjesson** [comment]
Hi,

I noticed a lot of code is reused from the DQN agent and the Dfdq agent implementations which will affect future maintainability of the code. If you could wrap the code that should be shared between the two implementations in methods, it would improve readability. 

Another thing I noticed was that it seems like the pretraining phase will run steps in the environment (among other things being that run during normal training phases) which seems unnecessary considering these steps aren't recorded into the memory. Avoiding this can substantially increase performance in some use cases.  My recommendation here would be to decouple this part from the rest of the implementation by wrapping the training of the network in its own method. This would also improve code readability and maintability as mentioned above. 

I haven't gone through all the code line by line but if I find the time I will try and verify that everything in your pull requests works as aspect. 

Great job so far though, I'm finding good use of the additions you've made in a personal project of mine.

File: rl/agents/dqfd.py [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    @property
+    def policy(self):
+        return self.__policy
+
+    @policy.setter
+    def policy(self, policy):
+        self.__policy = policy
+        self.__policy._set_agent(self)
+
+    @property
```
[the remark this answers is not in the document]
Comment (jakegrigsby) [reply rl/agents/dqfd.py]: I think this is copied straight from DQNAgent. DQfD inherits from AbstractDQNAgent so I thought it would need this as well?

File: examples/dqfd_atari.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+cv2 = Conv2D(64, kernel_size=(4,4), strides=2, activation='relu', kernel_regularizer=l2(1e-4), data_format='channels_first')(cv1)
+cv3 = Conv2D(64, kernel_size=(3,3), strides=1, activation='relu', kernel_regularizer=l2(1e-4), data_format='channels_first')(cv2)
+dense= Flatten()(cv3)
+dense = Dense(512, activation='relu', kernel_regularizer=l2(1e-4))(dense)
+buttons = Dense(nb_actions, activation='linear', kernel_regularizer=l2(1e-4))(dense)
+model = Model(inputs=frame,outputs=buttons)
+print(model.summary())
+
+processor = AtariDQfDProcessor()
+expert_demo_data = processor.process_demo_data(load_demo_data_from_file('gopher_expert_5k.npy'))
```
[the remark this answers is not in the document]
Comment (jakegrigsby) [reply examples/dqfd_atari.py]: I thought it would be too big to include. You can create it by uncommenting the line before it: `record_demo_data('GopherDeterministic-v4', steps=50000, data_filepath='gopher_expert.npy', frame_delay=0.03)`
(although that requires you to sit there and play the game for 50k steps).

I guess it comes down to whether we need the examples to be runnable out of the box, or if they can just serve as a template for people's own use cases.

File: rl/memory.py [hunk 3]
```diff
@@ -605,9 +605,7 @@ def _sample_proportional(self, batch_size):
 
         return idxs
 
-    def sample(self, batch_size, beta=1.):
-        idxs = self._sample_proportional(batch_size)
-
+    def sample(self, idxs, batch_size, beta=1., nstep=1, gamma=1):
```
[the remark this answers is not in the document]
Comment (jakegrigsby) [reply rl/memory.py]: I think you were looking at an outdated version. The Codacy auto code review suggested a change,  so I switched it to `sample_by_idxs`, leaving `sample` unimplemented.

File: examples/dqfd_atari.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jakegrigsby) [reply examples/dqfd_atari.py]: to clarify, that would be the line before it if you were looking at the most recent commit.

**Comment by jakegrigsby** [comment]
Thanks for the feedback. I went ahead and (hopefully) fixed most of the smaller mistakes caught by @RaphaelMeudec.

I'll work on adding tests and streamlining some of the duplicate dqn agent code. I think a good portion of it could be added to AbstractDQNAgent, which both dqn and dqfd could call.

I'm torn as far as the pretraining environment interaction goes. While it would be faster to avoid the environment during that phase (especially in bigger envs like Atari games), it does seem useful to be able to monitor the agent's progress/metrics from the beginning; the pretraining phase is typically where most of the learning occurs. Let me know what you think. I could probably do it by overriding the `fit` method of the `Agent` class so that `forward` isn't called before the pretraining steps are up. Then there'd probably need to be some changes to `backward` to keep it from trying to add transitions to memory.

**Comment by jakegrigsby** [comment]
Ended up switching it to env-free pretraining, didn't need too many changes. I managed to trim it down a little by sharing `forward` and the double q-learning algorithm between `DQNAgent` and `DQfDAgent`. Also added a test to test_dqn.py
